src/sync/app.cpp

```diff
@@ -873,13 +873,88 @@ void App::refresh_access_token(std::shared_ptr<SyncUser> sync_user,
+                        std::function<void (util::Optional<AppError>,
+                                            util::Optional<bson::Bson>)> completion_block)
+{
+    auto handler = [completion_block](const Response& response) {
+        if (auto error = check_for_errors(response)) {
+            return completion_block(error, util::none);
+        }
+
+        completion_block(util::none, util::Optional<bson::Bson>(bson::parse(response.body)));
+    };
+
+    std::string route = util::format("%1/app/%2/functions/call", m_base_route, m_config.app_id);
+
+    bson::BsonDocument args {
+        { "arguments", args_bson },
+        { "name", name }
+    };
+
+    if (service_name) {
+        args["service"] = *service_name;
+    }
+
+    std::stringstream s;
+    s << bson::Bson(args);
+
+    Request request {
+        .method = HttpMethod::post,
+        .url = route,
+        .body = s.str()
+    };
+
+    do_authenticated_request(request,
+                             user,
+                             handler);
+}
+
+void App::call_function(std::shared_ptr<SyncUser> user,
+                        const std::string& name,
+                        const bson::BsonArray& args_bson,
+                        std::function<void (util::Optional<AppError>,
+                                            util::Optional<bson::Bson>)> completion_block)
+{
+    call_function(user,
+                  name,
+                  args_bson,
+                  util::none,
+                  completion_block);
+}
+
+void App::call_function(const std::string& name,
+                        const bson::BsonArray& args_bson,
+                        const util::Optional<std::string>& service_name,
+                        std::function<void (util::Optional<AppError>,
+                                            util::Optional<bson::Bson>)> completion_block)
+{
+    call_function(SyncManager::shared().get_current_user(),
+                  name,
+                  args_bson,
+                  service_name,
+                  completion_block);
+}
+
+void App::call_function(const std::string& name,
+                        const bson::BsonArray& args_bson,
+                        std::function<void (util::Optional<AppError>,
+                                            util::Optional<bson::Bson>)> completion_block)
+{
+    call_function(SyncManager::shared().get_current_user(),
+                  name,
+                  args_bson,
+                  completion_block);
+}
 
 RemoteMongoClient App::remote_mongo_client(const std::string& service_name)
 {
-    return RemoteMongoClient(AppServiceClient(service_name,
-                                              m_base_route,
-                                              m_config.app_id,
-                                              *this));
+    // FIXME: Use shared_from_this once enabled
+    return RemoteMongoClient(std::static_pointer_cast<AppServiceClient>(std::make_shared<App>(m_config)),
```

Comment:
This looks wrong. It should be `SyncManager::shared().app()` Since the App is still a singleton (found this when debugging with @edualonso )

Not sure if this is done other places

Reply:
That's why the `FIXME` is here. Is it causing a crash on Java?

Reply:
Yes it is.

Reply:
I was hoping to avoid having to make this change in this PR, but I will fix this later today.